okhttp/src/jvmMain/kotlin/okhttp3/Dispatcher.kt

```diff
@@ -98,8 +99,17 @@ class Dispatcher() {
       return executorServiceOrNull!!
     }
 
+  /**
+   * Orders calls by order of execution. When two calls have equal priority order they are executed
+   * in the order that they were enqueued in.
+   */
+  @get:Synchronized
+  @set:Synchronized
+  @get:JvmName("priorityOrder")
+  var priorityOrder: Comparator<Call>? = null
+
   /** Ready async calls in the order they'll be run. */
-  private val readyAsyncCalls = ArrayDeque<AsyncCall>()
+  private val readyAsyncCalls = mutableListOf<AsyncCall>()
```

Comment:
The only nagging doubt is the cost of remove for ArrayList vs ArrayDeque.  ArrayDeque is more optimal for elements near the start, which is typical... 

Removing the first element is O(1) vs O(n)

Reply:
@yschimke I have to investigate but I've seen that sort function copies entire list to an array, sorts it, and then copies it back to original list.  
Maybe we can keep using array, but instrad of using iterator, we can use takeFirst() n and copy the rest of the array back.

Reply:
https://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/Collections.java#l164

Reply:
Yep - so it's better than a naive sort on a LinkedList, but it's still doing a lot more work when we consider we have a mostly sorted list already.

```
     * <p>The implementation was adapted from Tim Peters's list sort for Python
     * (<a href="http://svn.python.org/projects/python/trunk/Objects/listsort.txt">
     * TimSort</a>).  It uses techniques from Peter McIlroy's "Optimistic
     * Sorting and Information Theoretic Complexity", in Proceedings of the
     * Fourth Annual ACM-SIAM Symposium on Discrete Algorithms, pp 467-474,
     * January 1993.
     *
     * <p>This implementation dumps the specified list into an array, sorts
     * the array, and iterates over the list resetting each element
     * from the corresponding position in the array.  This avoids the
     * n<sup>2</sup> log(n) performance that would result from attempting
     * to sort a linked list in place.
```

Reply:
I'm not sure in practice whether this will be noticeable.  I wonder what performance testing square has, that would flush this out on Android and JVM.

If it's a concern and blocking, a possible way forward is the small domain model I suggested that would let us continue to use the current ArrayDeque when no priority comparator is set, and only take the hit when it's set.

```
 interface DispatchQueue {
  fun insert(call: AsyncCall)
  fun iterator(): Iterator<AsyncCall>
  fun sort()
}
```

But let's see if we are happy to land this for now, before changing anything.

Reply:
@yschimke maybe something like this?

https://github.com/tom5079/okhttp/commit/c5d7106a8274b2875d6b603f0bbe9f170dad0072